tests/trx_generator/trx_generator.cpp

```diff
@@ -202,4 +205,117 @@ namespace eosio::testing {
+         packed_action_data_string = abi.variant_to_binary( action_type, unpacked_action_data_json, abi_serializer::create_yield_function( abi_serializer_max_time ) );
+
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Fail to parse unpacked action data JSON")
+
+      std::cout << fc::to_hex(packed_action_data_string.data(), packed_action_data_string.size()) << std::endl;
+
+      eosio::chain::action act;
+      act.account = _contract_owner_account;
+      act.name = _action;
+      act.authorization = vector<permission_level>{{_auth_account, config::active_name}};
+      act.data = std::move(packed_action_data_string);
+
+      _trxs.emplace_back(create_transfer_trx_w_signer(act, _private_key, ++_nonce_prefix, _nonce, _trx_expiration, _chain_id, _last_irr_block_id));
+
+      std::cout << "Setup p2p transaction provider" << std::endl;
+
+      std::cout
+            << "Update each trx to qualify as unique and fresh timestamps, re-sign trx, and send each updated transactions via p2p transaction provider"
+            << std::endl;
+
+      _provider.setup();
+      return true;
+   }
+
+   bool trx_generator::tear_down() {
+      _provider.log_trxs(_log_dir);
+      _provider.teardown();
+
+      std::cout << "Sent transactions: " << _txcount << std::endl;
+      std::cout << "Tear down p2p transaction provider" << std::endl;
+
+      //Stop & Cleanup
+      std::cout << "Stop Generation." << std::endl;
+      stop_generation();
+      return true;
+   }
+
+   void trx_generator::push_transaction(p2p_trx_provider& provider, signed_transaction_w_signer& trx, uint64_t& nonce_prefix, uint64_t& nonce, const fc::microseconds& trx_expiration, const chain_id_type& chain_id, const block_id_type& last_irr_block_id) {
+      update_resign_transaction(trx._trx, trx._signer, ++nonce_prefix, nonce, trx_expiration, chain_id, last_irr_block_id);
+      provider.send(trx._trx);
+   }
+
+   bool trx_generator::generate_and_send() {
+      try {
+         if (_trxs.size()) {
+            size_t index_to_send = _txcount % _trxs.size();
+            push_transaction(_provider, _trxs.at(index_to_send), ++_nonce_prefix, _nonce, _trx_expiration, _chain_id,
+                             _last_irr_block_id);
+            ++_txcount;
+         } else {
+            elog("no transactions available to send");
+            return false;
+         }
+      } catch (const std::exception &e) {
+         elog("${e}", ("e", e.what()));
+         return false;
+      } catch (...) {
+         elog("unknown exception");
+         return false;
+      }
+
+      return true;
+   }
+
+   void trx_generator::stop_generation() {
+      ilog("Stopping transaction generation");
+
+      if(_txcount) {
+         ilog("${d} transactions executed, ${t}us / transaction", ("d", _txcount)("t", _total_us / (double) _txcount));
+         _txcount = _total_us = 0;
+      }
+   }
```

Comment:
`tear_down`, `push_transaction`, `generate_and_send`, and `stop_generation()` are all copy/paste. Seems like a base class is in order for some of this functionality boilerplate.

Reply:
Addressed: [03750c9](https://github.com/AntelopeIO/leap/pull/626/commits/03750c994aa61d4877ae708c2920245ff8e9526b)